lib/web/controller/mcp_server_helper.ex

```diff
@@ -0,0 +1,344 @@
+          param = arguments["param"] || "default"
+          McpServerHelper.response_text("Result: \#{param}")
+        end
+      end
+  """
+
+  # JSON-RPC 2.0 error codes
+  # https://www.jsonrpc.org/specification#error_object
+  @error_code_method_not_found -32_601
+  @error_code_internal_error -32_603
+  @error_code_server_error -32_000
+
+  defmodule ToolCallback do
+    @moduledoc """
+    Type definition for a tool callback function.
+
+    A callback function takes a connection and arguments map, and returns a result map.
+    """
+
+    @type t :: (Antikythera.Conn.t(), map() -> map())
+
+    defun valid?(f :: any) :: boolean do
+      is_function(f, 2)
+    end
+
+    defun default() :: nil do
+      nil
+    end
+  end
+
+  defmodule Tool do
+    @moduledoc """
+    Defines a tool that can be called via MCP (Model Context Protocol).
+
+    ## Fields
+
+    - `name` (required, string) - Unique identifier for the tool
+    - `description` (required, string) - Human-readable description of what the tool does
+    - `inputSchema` (required, map) - JSON Schema defining the input parameters
+    - `outputSchema` (optional, map) - JSON Schema defining the output format
+    - `callback` (required, function) - Function that handles the tool call, signature: `(conn, arguments) -> map`
+
+    ## Example
+
+        McpServerHelper.Tool.new!(%{
+          name: "echo",
+          description: "Echoes back the input text",
+          inputSchema: %{
+            type: "object",
+            properties: %{
+              text: %{type: "string", description: "Text to echo"}
+            },
+            required: ["text"]
+          },
+          callback: &MyController.handle_echo/2
+        })
+    """
+
+    use Croma.Struct,
+      fields: [
+        name: Croma.String,
+        description: Croma.String,
+        inputSchema: Croma.Map,
+        outputSchema: Croma.TypeGen.nilable(Croma.Map),
+        callback: ToolCallback
+      ]
+  end
+
+  defmacro __using__(opts) do
+    server_name = Keyword.fetch!(opts, :server_name)
+    server_version = Keyword.fetch!(opts, :server_version)
+    tools = Keyword.fetch!(opts, :tools)
+
+    quote do
+      alias Antikythera.{Conn, G2gResponse}
+
+      @mcp_server_name unquote(server_name)
+      @mcp_server_version unquote(server_version)
+      @mcp_tools unquote(tools)
+
+      def handle_mcp_request(conn) do
+        request_body = conn.request.body
+
+        case request_body do
+          %{"method" => "initialize"} ->
+            handle_mcp_initialize(conn, request_body)
+
+          %{"method" => "notifications/initialized"} ->
+            handle_mcp_notification(conn)
+
+          %{"method" => "tools/list"} ->
+            handle_mcp_tools_list(conn, request_body)
+
+          %{"method" => "tools/call"} ->
+            handle_mcp_tools_call(conn, request_body)
+
+          _ ->
+            Conn.json(conn, 400, %{error: "Unknown method"})
+        end
+      end
```

Comment:
Will the batch RPC calls be supported? Clients may send an array batching one or more request objects.
**Reference:** <https://modelcontextprotocol.io/specification/2025-03-26/basic/transports#sending-messages-to-the-server>